nautobot/core/api/serializers.py

```diff
@@ -132,14 +216,107 @@ def get_field_names(self, declared_fields, info):
         fields = list(super().get_field_names(declared_fields, info))  # Meta.fields could be defined as a tuple
         self.extend_field_names(fields, "display", at_start=True)
         self.extend_field_names(fields, "id", at_start=True)
+        self.eliminate_field_names(fields, "_custom_field_data")
+        self.eliminate_field_names(fields, "_name")
         # Needed because we don't have a common base class for all nested serializers vs non-nested serializers
         if not self.__class__.__name__.startswith("Nested"):
             if hasattr(self.Meta.model, "created"):
                 self.extend_field_names(fields, "created")
             if hasattr(self.Meta.model, "last_updated"):
                 self.extend_field_names(fields, "last_updated")
+        # append non-default model api fields to display in Nautobot API
+        # e.g. for annotated fields `circuit_count`, `device_count` and etc.
+        if getattr(self.Meta, "extra_fields", None):
+            return fields + self.Meta.extra_fields
         return fields
 
+    def to_internal_value(self, data):
+        for key, value in self.get_fields().items():
+            if value.__class__.__name__ in ["NautobotNestedSerializer", "NautobotPrimaryKeyRelatedField"]:
+                sub_data = data.get(key, None)
+                if sub_data is not None:
+                    data[key] = value.to_internal_value(sub_data)
+        return data
+
+    def build_nested_field(self, field_name, relation_info, nested_depth):
+        field = get_serializer_for_model(relation_info.related_model)
+
+        class NautobotNestedSerializer(field):
```

Comment:
This class logic looks really similar to `NautobotPrimaryKeyRelatedField`  above, I guess finding a way to utitlize that class here might be a better approach